src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/SingleInitNullable.cs

```diff
@@ -0,0 +1,74 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using Roslyn.Utilities;
+
+/// <summary>
+/// A lazily initialized version of <see cref="Nullable{T}"/> which uses the same space as a <see cref="Nullable{T}"/>.
+/// </summary>
+[NonCopyable]
+internal struct SingleInitNullable<T>
+    where T : struct
+{
+    /// <summary>
+    /// One of three values:
+    /// <list type="bullet">
+    /// <item>0. <see cref="_value"/> is not initialized yet.</item>
+    /// <item>1. <see cref="_value"/> is currently being initialized by some thread.</item>
+    /// <item>2. <see cref="_value"/> has been initialized.</item>
+    /// </list>
+    /// </summary>
+    private int _initialized;
+
+    /// <summary>
+    /// Actual stored value.  Only safe to read once <see cref="_initialized"/> is set to 2.
+    /// </summary>
+    private T _value;
+
+    /// <summary>
+    /// Ensure that the given target value is initialized in a thread-safe manner.
+    /// </summary>
+    /// <param name="valueFactory">A factory delegate to create a new instance of the target value. Note that this
+    /// delegate may be called more than once by multiple threads, but only one of those values will successfully be
+    /// written to the target.</param>
+    /// <returns>The target value.</returns>
+    public T Initialize<TArg>(Func<TArg, T> valueFactory, TArg arg)
+        => ReadIfInitialized() ?? GetOrStore(valueFactory(arg));
```

Comment:
note: we could push valueFactory into GetOrStore. That would ensure that it was only called once. But it could mean another thread might then spin a long time while the computation was active.

Reply:
I admit that was what I expected state 1 to be: otherwise we're only ensuring there isn't multiple initialization if different sides computed different values, which obviously isn't a concern in the checksum case.

Reply:
Yeah it wasn't until after I finding that reset logic (and not finding it!) I realized it wasn't doing what I expected. :smile: